src/events/attributes.rs

```diff
@@ -37,39 +37,33 @@ impl<'a> Attribute<'a> {
     ///
     /// This will allocate if the value contains any escape sequences.
     ///
-    /// See also [`unescaped_value_with_custom_entities()`](Self::unescaped_value_with_custom_entities)
+    /// See also [`unescaped_value_with()`](Self::unescaped_value_with)
     pub fn unescaped_value(&self) -> XmlResult<Cow<[u8]>> {
-        self.make_unescaped_value(None)
+        self.unescaped_value_with(|_| None)
     }
 
     /// Returns the unescaped value, using custom entities.
     ///
     /// This is normally the value you are interested in. Escape sequences such as `&gt;` are
     /// replaced with their unescaped equivalents such as `>`.
-    /// Additional entities can be provided in `custom_entities`.
+    /// A fallback resolver for additional custom entities can be provided via
+    /// `resolve_entities`.
     ///
     /// This will allocate if the value contains any escape sequences.
     ///
     /// See also [`unescaped_value()`](Self::unescaped_value)
     ///
     /// # Pre-condition
     ///
-    /// The keys and values of `custom_entities`, if any, must be valid UTF-8.
-    pub fn unescaped_value_with_custom_entities(
-        &self,
-        custom_entities: &HashMap<Vec<u8>, Vec<u8>>,
-    ) -> XmlResult<Cow<[u8]>> {
-        self.make_unescaped_value(Some(custom_entities))
+    /// The implementation of `resolve_entity` is expected to operate over UTF-8 inputs.
+    pub fn unescaped_value_with<'s, 'entity>(
+        &'s self,
+        resolve_entity: impl Fn(&[u8]) -> Option<&'entity str>,
+    ) -> XmlResult<Cow<'s, [u8]>> {
+        unescape_with(&*self.value, resolve_entity).map_err(Error::EscapeError)
```

Comment:
I think it will be more consistently with other codebase

```suggestion
    #[cfg(not(feature = "encoding"))]
    pub fn unescape_value_with<'s, 'entity>(
        &'s self,
        resolve_entity: impl Fn(&[u8]) -> Option<&'entity str>,
    ) -> XmlResult<Cow<'s, [u8]>> {
        Ok(unescape_with(&*self.value, resolve_entity)?)
```

The added feature gate prevents incorrect usage. All usages in tests should be replaced with `decode_*` variant. Another way to solve that -- add a decoder to the `Attribute` / `Event` and keep only variant without `reader` parameter, but I think that this may be a task for another PR, focused on encoding.

Reply:
I would rather push all the decoding changes into a different PR.  At the moment all of the parsing code is based on searches for single ascii bytes, which means it is effectively a UTF-8 only library, even if the detection of encodings works properly.  https://github.com/tafia/quick-xml/issues/322

So there's a lot of work needed to make the `encoding` feature actually useful.

Reply:
Yes, I understand and support. That is why I suggest just disable dangerous methods when `"encoding"` feature is enabled.

> So there's a lot of work needed to make the `encoding` feature actually useful.

Actually, it is useful for one-byte encodings. Bearing in mind, that
- we use `encoding_rs`
- [`encoding_rs` is not extensible by design](https://docs.rs/encoding_rs/latest/encoding_rs/index.html#no-extensibility-by-design)
- as the result, the list of encodings is fixed
- the only encodings, that is not XML compatible (because is not ASCII compatible, which is more strict restriction) is (generated by [this snippet](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=28602c63d99403b1edb96db8c615875d)):
  - `ISO-2022-JP`
  - `replacement` (is not a real encoding, actually)
  - `UTF-16BE`
  - `UTF-16LE`

So the `encoding` feature actually may be useful even in the current state.

----
The other suggestion is to make namings is even more consistently, compare:
- `unescaped`
- `decode_and_unescape`

Because the difference only in the `decode` operation applied in the second function, it is logically, that stripping out the `decode_and_` prefix we should get a name of the method that works without decoding

Reply:
You are right about the other one-byte encodings, I just don't know how common they actually are, certainly compared to UTF-16.  But probably what I should have said is that I don't really want to go back and forth over functions that it might make sense to remove entirely at some point in the future.

Reply:
I will go ahead and make the changes anyway (tomorrow, since I really need to go to sleep now)

Reply:
@Mingun I wont forget about this, but I am going to put this in a separate PR.  I want to finish running some experiments first re: bulk decode vs. decoding individual elements and the outcome of that might influence what we want the API to look like.

Reply:
I'm totally agree that encoding part should be in a separate PR, but can you polish namings in this PR? Only one simple change:
- `unescaped()` -> `unescape()`
- `unescaped_with()` -> `unescape_with()`
- `unescaped_value()` -> `unescape_value()`
- `unescaped_value_with()` -> `unescape_value_with()`

As you already changed some of that names it is quite logically to finish with that here, just to keep history clean :). And after that it can be merged

Reply:
OK :)  but, please take a look at https://github.com/tafia/quick-xml/issues/158#issuecomment-1185081402

I'm going to draft this for now because I'm 100% convinced we should change the APIs to pure `&str` / `String` ones.

Reply:
> I'm going to draft this for now because I'm 100% convinced we should change the APIs to pure `&str` / `String` ones.

I agree, I already work in this direction on top of this PR, namely changing the order of unescape and decode operations in the serde deserializer. I think I'm going to merge it now and continue evolving API in the another PRs. At least I want to make the new closure-based API accepts `&str` instead of `&[u8]` before the release.